README.md
# NAME

Carl - yet another Carton

# SYNOPSIS

Development time:

    > carl install
    ...
    Complete! Modules were installed to local, and wrote index.txt

    > git add index.txt
    > git commit -m 'add index.txt'

Deployment time, only cpanm!

    > cpanm --mirror-index $PWD/index.txt -nq -Llocal --installdeps .

# DESCRIPTION

Carl is yet another Carton.

# FEATURES

- (1) use index.txt (02packages.details.txt) to save versions
- (2) so, when deployment, no need to install Carl itself, just cpanm
- (3) support git/dist dependencies

Let me explain (3).
Suppose you have the following cpanfile:

    requires 'My::Module', git => 'git://ghe.example.com/you/My-Module.git', ref => '0.01';
    requires 'Anthor::Module', dist => 'http://darkpan.example.com/Anothor-Module-0.01.tar.gz';

Carl supports it!

    > carl install
    ...
    Complete! Modules were installed to local, and wrote index.txt

    # I recommend you to git-add local/cache
    > git add index.txt local/cache
    > git commit -m 'add index.txt and local/cache for consistent installation'

    # deployment time!
    > cpanm --mirror-index $PWD/index.txt --mirror file://$PWD/local/cache -nq -Llocal --installdeps .

# SEE ALSO

Much code or architecture are stolen from Carton and Carmel.

[Carton](https://metacpan.org/pod/Carton)

[https://github.com/miyagawa/Carmel](https://github.com/miyagawa/Carmel)

[App::cpanminus](https://metacpan.org/pod/App::cpanminus)

[https://github.com/miyagawa/cpan-module-bootstrap](https://github.com/miyagawa/cpan-module-bootstrap)

[OrePAN2](https://metacpan.org/pod/OrePAN2)

[cpanfile](https://metacpan.org/pod/cpanfile)

[https://speakerdeck.com/miyagawa/whats-new-in-carton-and-cpanm-at-yapc-asia-2013](https://speakerdeck.com/miyagawa/whats-new-in-carton-and-cpanm-at-yapc-asia-2013)

# LICENSE

Copyright (C) Shoichi Kaji.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

# AUTHOR

Shoichi Kaji <skaji@cpan.org>


Changes
Revision history for Perl extension Carl

{{$NEXT}}

    - original version



lib/Carl/CLI.pm
package Carl::CLI;
use strict;
use warnings;
use Carl::Builder;
use Carl::Indexer;
use Carl::LocalMirror;
use Config;
use Cwd ();
use File::Temp ();

sub new {
    my ($class, %option) = @_;
    bless { %option }, $class;
}

sub run {
    my ($self, @argv) = @_;
    my $cmd = shift @argv || "help";
    if ( my $code = $self->can( "cmd_$cmd" ) ) {
        $self->$code(@argv);
    } else {
        warn "Unknown command $cmd.\n";
        $self->cmd_help;
    }
}

sub cmd_help {
    exec "perldoc", "Carl";
    exit 255;
}

sub cmd_install {
    my $self = shift;
    my $directory = File::Temp::tempdir(CLEANUP => !$self->{debug});

    my $mirror = Carl::LocalMirror->new(
        directory => $directory,
        cpanfile => "cpanfile",
        debug => $self->{debug},
    );
    my $setuped = $mirror->setup;

    my $builder = Carl::Builder->new(
        ( $setuped ? (mirror => $directory) : () ),
        directory => "local",
        cpanfile => "cpanfile",
    );
    $builder->install;

    my $archlib = "local/lib/perl5/$Config{archname}";
    my $indexer = Carl::Indexer->new(directory => $archlib);
    open my $fh, ">", "index.txt" or die "Cannot open index.txt: $!";
    $indexer->write_index($fh);

    warn "Complete! Modules were installed to local, and wrote index.txt\n";
}

sub cmd_exec {
    my ($self, @argv) = @_;
    my $cwd = Cwd::cwd;
    $ENV{PERL5LIB} = ($ENV{PERL5LIB} ? "$ENV{PERL5LIB}:" : "") . "$cwd/local/lib/perl5";
    $ENV{PATH} = "$cwd/local/bin:$ENV{PATH}";
    exec @argv;
    exit 255;
}

1;


lib/Carl/Builder.pm
package Carl::Builder;
use strict;
use warnings;
use Module::Metadata;

sub new {
    my ($class, %option) = @_;
    bless {%option}, $class;
}

sub install {
    my $self = shift;
    my @option;
    if (my $mirror = $self->{mirror}) {
        push @option,
            "--cascade-search",
            "--mirror" => "file://$mirror",
            "--mirror-index" => "$mirror/modules/02packages.details.txt",
            "--mirror" => "http://www.cpan.org";
    }
    $self->run_cpanm(
        @option,
        "--save-dists" => "$self->{directory}/cache",
        "-L", $self->{directory},
        "--cpanfile", $self->{cpanfile},
        "--installdeps", ".",
    ) or die "Install failed\n";
}

sub run_cpanm {
    my ($self, @argv) = @_;
    local $ENV{PERL_CPANM_OPT};
    my $fatscript = Module::Metadata->find_module_by_name("App::cpanminus::fatscript");
    !system $^X, $fatscript, "-nq", @argv;
}

1;


lib/Carl/Indexer.pm
package Carl::Indexer;
use strict;
use warnings;
use Config;
use File::Find 'find';
use JSON::PP;
use List::Util 'sum';
use Carl;

our $VERSION = $Carl::VERSION;

sub new {
    my ($class, %option) = @_;
    my $directory = delete $option{directory} or die;
    bless { %option, directory => $directory }, $class;
}

# header is copied from Carmel::App::write_index
sub _write_index {
    my ($self, $fh, @distribution) = @_;

    my $count = sum(0, map { scalar(@{$_->{package}}) } @distribution) + 1;

    print $fh <<EOF;
File:         02packages.details.txt
URL:          http://www.perl.com/CPAN/modules/02packages.details.txt
Description:  Package names found in local directories
Columns:      package name, version, path
Intended-For: Automated fetch routines, namespace documentation.
Written-By:   @{[ sprintf "%s %s", __PACKAGE__, __PACKAGE__->VERSION ]}
Line-Count:   $count

EOF

    for my $dist (@distribution) {
        my $pathname = $dist->{pathname};
        for my $package (@{ $dist->{package} }) {
            printf $fh "%-36s %-8s %s\n",
                $package->{name}, $package->{version} // 'undef', $pathname;
        }
    }
}

sub write_index {
    my ($self, $fh) = @_;
    my @install_json = $self->gather_install_json;
    my @distribution;
    for my $json (@install_json) {
        my $meta = decode_json do { local (@ARGV, $/) = $json; <> };
        push @distribution, {
            pathname => $meta->{pathname},
            package => [
                map +{
                    name => $_,
                    version => $meta->{provides}{$_}{version},
                }, sort keys %{$meta->{provides}}
            ],
        };
    }
    $self->_write_index($fh, @distribution);
}

sub gather_install_json {
    my $self = shift;
    my @install_json;
    find sub {
        return unless -f $_ && $_ eq "install.json";
        push @install_json, $File::Find::name;
    }, $self->{directory};
    sort @install_json;
}

1;


lib/Carl/LocalMirror.pm
package Carl::LocalMirror;
use 5.008001;
use strict;
use warnings;
use OrePAN2::Injector;
use OrePAN2::Indexer;
use Module::CPANfile;
use Capture::Tiny 'capture_merged';

sub new {
    my ($class, %option) = @_;
    my $directory = delete $option{directory} or die;
    my $cpanfile  = delete $option{cpanfile}  or die;
    bless { %option, cpanfile => $cpanfile, directory => $directory }, $class;
}

sub debug {
    my $self = shift;
    return unless $self->{debug};
    warn "--> @_\n";
}

sub setup {
    my $self = shift;

    my $cpanfile = Module::CPANfile->load($self->{cpanfile});
    my @module = $cpanfile->merged_requirements->required_modules;

    my @need_inject;
    for my $module (@module) {
        my $option = $cpanfile->options_for_module($module) || +{};
        if (my $git = $option->{git}) {
            $git = "$git\@$option->{ref}" if $option->{ref};
            push @need_inject, { module => $module, from => $git };
        } elsif (my $dist = $option->{dist}) {
            # which dist support? read OrePAN2::Injector :-)
            push @need_inject, { module => $module, from => $dist };
        }
    }
    return unless @need_inject;

    my $directory = $self->{directory};
    $self->debug("building local repository $directory");
    my $injector = OrePAN2::Injector->new(directory => $directory);
    for my $module (@need_inject) {
        $self->debug("injecting $module->{module} from $module->{from}");
        my $merged = capture_merged { $injector->inject($module->{from}) };
    }
    my $indexer = OrePAN2::Indexer->new(directory => $directory);
    $self->debug("indexing $directory");
    my $merged = capture_merged { $indexer->make_index( no_compress => 1 ) };
    return 1;
}

1;


lib/Carl.pm
package Carl;
use 5.008001;
use strict;
use warnings;

our $VERSION = "0.01";
1;
__END__

=encoding utf-8

=head1 NAME

Carl - yet another Carton

=head1 SYNOPSIS

Development time:

    > carl install
    ...
    Complete! Modules were installed to local, and wrote index.txt

    > git add index.txt
    > git commit -m 'add index.txt'

Deployment time, only cpanm!

    > cpanm --mirror-index $PWD/index.txt -nq -Llocal --installdeps .

=head1 DESCRIPTION

Carl is yet another Carton.

=head1 FEATURES

=over 4

=item (1) use index.txt (02packages.details.txt) to save versions

=item (2) so, when deployment, no need to install Carl itself, just cpanm

=item (3) support git/dist dependencies

=back

Let me explain (3).
Suppose you have the following cpanfile:

    requires 'My::Module', git => 'git://ghe.example.com/you/My-Module.git', ref => '0.01';
    requires 'Anthor::Module', dist => 'http://darkpan.example.com/Anothor-Module-0.01.tar.gz';

Carl supports it!

    > carl install
    ...
    Complete! Modules were installed to local, and wrote index.txt

    # I recommend you to git-add local/cache
    > git add index.txt local/cache
    > git commit -m 'add index.txt and local/cache for consistent installation'

    # deployment time!
    > cpanm --mirror-index $PWD/index.txt --mirror file://$PWD/local/cache -nq -Llocal --installdeps .

=head1 SEE ALSO

Much code or architecture are stolen from Carton and Carmel.

L<Carton>

L<https://github.com/miyagawa/Carmel>

L<App::cpanminus>

L<https://github.com/miyagawa/cpan-module-bootstrap>

L<OrePAN2>

L<cpanfile>

L<https://speakerdeck.com/miyagawa/whats-new-in-carton-and-cpanm-at-yapc-asia-2013>

=head1 LICENSE

Copyright (C) Shoichi Kaji.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

Shoichi Kaji E<lt>skaji@cpan.orgE<gt>

=cut

